proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,754 @@
+* `map` would still need to be resolved via overload resolution amongst the existing overloads, so this approach doesn't help much with the type checking issues that `...` has.
+
+## Future directions
+
+### Variadic generic types
+
+This proposal only supports type parameter packs on functions. A complementary proposal will describe type parameter packs on generic structs, enums and classes.
+
+### Value expansion operator
+
+This proposal only supports the expansion operator on type parameter packs and value parameter packs, but there are other values that represent a list of zero or more values that the expansion operator would be useful for, including tuples and arrays. It would be desirable to introduce a new kind of expression that receives a scalar value and produces a value of pack expansion type.
+
+Here, we use the straw-man syntax `.element` for accessing tuple elements as a pack:
+
+```swift
+func foo<each T, each U>(
+  _ t: repeat each T,
+  _ u: repeat each U
+) {}
+
+func bar1<each T, each U>(
+  t: (repeat each T),
+  u: (repeat each U)
+) {
+  repeat foo(each t.element, each u.element)
+}
+
+func bar2<each T, each U>(
+  t: (repeat each T),
+  u: (repeat each U)
+) {
+  repeat foo(each t.element, repeat each u.element)
+}
+```
+
+Here, `bar1(t: (1, 2), u: ("a", "b"))` will evaluate:
+
+```swift
+foo(1, "a")
+foo(2, "b")
+```
+
+While `bar2(t: (1, 2), u: ("a", "b"))` will evaluate:
+
+```swift
+foo(1, "a", "b")
+foo(2, "a", "b")
+```
+
+The distinction can be understood in terms of our notion of _captures_ in pack expansion expressions.
+
+### Pack destructuring operations
+
+The statically-known shape of a pack can enable destructing packs with concrete shape into the component elements:
+
+```swift
+struct List<each Element> {
+  let element: repeat each Element
+}
+
+extension List {
+  func firstRemoved<First, each Rest>() -> List<repeat each Rest> where (repeat each Element) == (First, repeat each Rest) {
+    let (first, rest) = (repeat each element)
+    return List(repeat each rest)
+  }
+}
+
+let list = List(1, "Hello", true)
+let firstRemoved = list.firstRemoved() // 'List("Hello", true)'
+```
+
+The body of `firstRemoved` decomposes `Element` into the components of its shape -- one value of type `First` and a value pack of type `repeat each Rest` -- effectively removing the first element from the list.
+
+### Tuple conformances
+
+Parameter packs, the above future directions, and a syntax for declaring tuple conformances based on [parameterized extensions](https://github.com/apple/swift/blob/main/docs/GenericsManifesto.md#parameterized-extensions) over non-nominal types enable implementing custom tuple conformances:
+
+```swift
+extension<each T: Equatable> (repeat each T): Equatable {
+  public static func ==(lhs: Self, rhs: Self) -> Bool {
+    for (l, r) in repeat (each lhs.element, each rhs.element) {
+      guard l == r else { return false }
+    }
+    return true
+  }
+}
+```
+
```

Comment:
Length constraints were a huge part of the initial discussion thread on the pitch.

This proposal very clearly above explains why arbitrary shape requirements are not desirable. Earlier, it also states that any explicit syntax isn't proposed _for now_ because it shouldn't be necessary given implicit inference: would that make such an explicit syntax a _future direction_, or is it an _alternative considered_ (no longer desirable to pursue even as a future direction)?

In the first pitch thread, there was also commentary on the usefulness of "one-or-more-elements" packs versus just the "zero-or-more-elements" packs: even if no explicit syntax for arbitrary element count is being pursued, some might argue that "one-or-more" is a sufficiently useful special case and wonder if that's something that can or should be pursued, I suppose?

Reply:
> Earlier, it also states that any explicit syntax isn't proposed for now because it shouldn't be necessary given implicit inference: would that make such an explicit syntax a future direction, or is it an alternative considered

I added a comment to the proposal that an explicit same-shape requirement syntax is a future direction.

> In the first pitch thread, there was also commentary on the usefulness of "one-or-more-elements" packs versus just the "zero-or-more-elements" packs: even if no explicit syntax for arbitrary element count is being pursued, some might argue that "one-or-more" is a sufficiently useful special case and wonder if that's something that can or should be pursued

This direction is just sugar for a parameter pack that is known to have a partially concrete shape where the first element is a scalar type parameter, and the remainder of the shape is abstract. This is conceptually similar to something you can already write under this proposal:

```swift
func requiresOne<First, each T>(values first: First, _ other: repeat each T) {
  // do something with {first, repeat each other}
}
```

I don't know that it's possible to _only_ allow this form of partially concrete shapes without considering them in their full generality, and the future directions section is already pretty extensive for a single proposal, so I'm not sure it's worth calling this out specifically (and I personally would rather fully implement partially concrete shapes). We'll be prepared to answer that question on the review thread, though. Does that sound okay?

Reply:
Makes sense to me!